### PR TITLE
Custom Error Handling Feature

### PR DESCRIPTION
**Pull request by phrmsilva**
Adds the ability for the user to specify how they want to handle specific errors.

### CODE REVIEW

**Comment by phrmsilva** [comment]
@ranierorusso @morganjbruce @dnlserrano 
Please review this PR.

File: spec/breaker_spec.rb [hunk 1]
```diff
@@ -52,6 +52,71 @@
...
+          redis: MockRedis.new,
+          ignore: [RuntimeError, TypeError],
+          error_handler: Proc.new { |e, &h|
+            if (e.is_a?(ZeroDivisionError))
+              h.call(e)
+            else
+              if (e.is_a?(RuntimeError) or e.is_a?(IndexError))
+                raise e
+              else
+                42
```
Comment (ranierorusso) [inline spec/breaker_spec.rb:72]: Is this a Douglas Adams reference???

**Comment by ranierorusso** [comment]
My recommendation is to shift the focus of this change since what we are really trying to do is refine the 'ignore' option. With the current version (in master) we can ignore errors based just on their type but what we need is the ability to inspect the error object's attributes in order to decide whether to ignore it (e.g. we want to handle Savon::HTTPError differently depending on whether it's a 5xx error or a 4xx error). 

The PR offers a way to do this, but my concern is that in not being directed towards and restricted to this objective, it offers more flexibility than necessary. For example if you have RuntimeError in the ignore list, you can still negate that and have it be processed within the error_handler block. You can also have the error_handler block not raise an error at all; a feature which I see as being more a responsibility of the code that calls CB2 rather than CB2 itself, which should just decide whether to process an error or not, and then raise it. 

One alternative approach is the one taken in the paper_trail gem which allows model object attributes to be audited based on name but also allows passing a Proc which returns true or false in order to allow more control.

https://github.com/airblade/paper_trail/#2c-choosing-attributes-to-monitor
https://github.com/airblade/paper_trail/blob/master/lib/paper_trail/record_trail.rb ln 135

@morganjbruce, @dnlserrano, what do you think?

**Comment by dnlserrano** [comment]
@phrmsilva: Great job! 🙌  I am happy with the current solution and I would suggest it can be merged in as is, if it is urgent for somebody. I agree, however, that we would benefit from a cleaner way of setting up the "decision procs", if you will. This may have consequences in terms of the major version, though. Not sure.

@ranierorusso: Based on the `paper_trail` examples you pointed out, are you implicitly suggesting something like the following:

``` ruby
cb = CB2::Breaker.new(
  service: "aws",
  ...
  error_handler: {
    RuntimeError      => Proc.new { |e, &h| ... },
    ZeroDivisionError => Proc.new { |e, &h| ... },
    ...
  }
)
```

**Comment by ranierorusso** [comment]
@dnlserrano if you take a look at the PR @phrmsilva has subsequently created (https://github.com/onfido/cb2/pull/3), you'll see an example of the implementation, which is more just:

```
handle: {
    RuntimeError        => Proc.new { |e| e.message == "foo" },
    ZeroDivisionError => Proc.new { |e| e.message != "bar" },
    ...
  }
```

where the proc returns true or false to indicate if the error should be handled or ignored.

**Comment by phrmsilva** [comment]
Unless there are any strong objections from Morgan, we'll go with the strategy in https://github.com/onfido/cb2/pull/3 instead.

**Comment by phrmsilva** [comment]
Merged #3 instead.

**Pull request closed without merging**